cdap-app-fabric/src/main/java/co/cask/cdap/metadata/FieldLineageAdmin.java

```diff
@@ -0,0 +1,308 @@
+    }
+    return new FieldLineageDetails(incoming, outgoing);
+  }
+
+  private List<ProgramFieldOperationInfo> processOperations(Set<ProgramRunOperations> programRunOperations) {
+    List<ProgramFieldOperationInfo> result = new ArrayList<>();
+    for (ProgramRunOperations entry : programRunOperations) {
+      List<ProgramInfo> programInfo = computeProgramInfo(entry.getProgramRunIds());
+      List<FieldOperationInfo> fieldOperationInfo = computeFieldOperationInfo(entry.getOperations());
+      result.add(new ProgramFieldOperationInfo(programInfo, fieldOperationInfo));
+    }
+    return result;
+  }
+
+  /**
+   * Computes the list of {@link ProgramInfo} from given set of ProgramRunIds.
+   * For each program, there is only one entry in the returned list and it is sorted
+   * by the last executed time in descending order.
+   * @param programRunIds set of program run ids from which program info to be computed
+   * @return list of ProgramInfo
+   */
+  private List<ProgramInfo> computeProgramInfo(Set<ProgramRunId> programRunIds) {
+    Map<ProgramId, Long> programIdToLastExecutedTime = new HashMap<>();
+    for (ProgramRunId programRunId : programRunIds) {
+      long programRunExecutedTime = RunIds.getTime(programRunId.getRun(), TimeUnit.SECONDS);
+      Long lastExecutedTime = programIdToLastExecutedTime.get(programRunId.getParent());
+      if (lastExecutedTime == null || programRunExecutedTime > lastExecutedTime) {
+        programIdToLastExecutedTime.put(programRunId.getParent(), programRunExecutedTime);
+      }
+    }
+
+    Stream<Map.Entry<ProgramId, Long>> sortedByLastExecutedTime
+            = programIdToLastExecutedTime.entrySet().stream()
+                                          .sorted(Collections.reverseOrder(Map.Entry.comparingByValue()));
+
+    List<ProgramInfo> programInfos = new ArrayList<>();
+    sortedByLastExecutedTime.forEachOrdered(programIdLongEntry
+            -> programInfos.add(new ProgramInfo(programIdLongEntry.getKey(), programIdLongEntry.getValue())));
+
+    return programInfos;
+  }
+
+  /**
+   * Computes list of {@link FieldOperationInfo} from the given operations.
+   * Returned list contains the operations sorted in topological order i.e. each operation
+   * in the list is guaranteed to occur before any other operation which reads its outputs.
+   * @param operations set of operation to convert to FieldOperationInfo instances
+   * @return list of FieldOperationInfo sorted topologically
+   */
+  private List<FieldOperationInfo> computeFieldOperationInfo(Set<Operation> operations) {
+    Set<String> writeOperations = new HashSet<>();
+    Map<String, Operation> operationsMap = new HashMap<>();
+    for (Operation operation : operations) {
+      operationsMap.put(operation.getName(), operation);
+      if (operation.getType().equals(OperationType.WRITE)) {
+        writeOperations.add(operation.getName());
+      }
+    }
+
+    LinkedHashSet<String> orderedOperations = new LinkedHashSet<>();
```

Comment:
is this supposed to produce topological order? I am not sure it is correct. Consider:
```
a--+-------------+--d
    \           /
     +--b---c--+
```

Reply:
fixed the code and added test case for this scenario.